File: NET_4.5.2/C1.WPF.FlexGrid/VB/ExcelGrid/ExcelGrid/FlexExcel/ExcelCellFactory.vb
Imports System.Text
Imports C1.WPF.FlexGrid

Namespace ExcelGrid
	''' <summary>
	''' Custom cell factory that displays Excel-style row and column headers.
	''' </summary>
	Public Class ExcelCellFactory
		Inherits CellFactory
		Public Overrides Sub CreateTopLeftContent(ByVal grid As C1FlexGrid, ByVal bdr As Border, ByVal range As CellRange)
			Dim poly = CreatePolygon(0, 10, 10, 10, 10, 0)
			poly.Fill = New SolidColorBrush(Color.FromArgb(&He0, &Hf5, &Hf5, &Hf5))
			poly.VerticalAlignment = VerticalAlignment.Center
			poly.HorizontalAlignment = HorizontalAlignment.Right
			poly.Margin = New Thickness(2)
			bdr.Child = poly
			bdr.Tag = grid
			AddHandler bdr.MouseLeftButtonDown, AddressOf bdr_MouseLeftButtonDown
		End Sub
		Private Sub bdr_MouseLeftButtonDown(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			' select the whole grid when user clicks the top left cell
			Dim g = TryCast((CType(sender, FrameworkElement)).Tag, C1FlexGrid)
			g.SelectAll()
		End Sub
		Public Overrides Sub CreateRowHeaderContent(ByVal grid As C1FlexGrid, ByVal bdr As Border, ByVal range As CellRange)
			Dim tb = New TextBlock()
			tb.HorizontalAlignment = HorizontalAlignment.Center
			tb.VerticalAlignment = VerticalAlignment.Center
			tb.Text = String.Format("{0}", range.Row + 1)
			bdr.Child = tb
		End Sub
		Public Overrides Sub CreateColumnHeaderContent(ByVal grid As C1FlexGrid, ByVal bdr As Border, ByVal range As CellRange)
			Dim tb = New TextBlock()
			tb.HorizontalAlignment = HorizontalAlignment.Center
			tb.VerticalAlignment = VerticalAlignment.Center
			tb.Text = GetAlphaColumnHeader(range.Column)
			bdr.Child = tb
		End Sub

		' create Excel style column headers ("A", "B", ... "AB", etc)
		Private Function GetAlphaColumnHeader(ByVal i As Integer) As String
			Return GetAlphaColumnHeader(i, String.Empty)
		End Function
		Private Function GetAlphaColumnHeader(ByVal i As Integer, ByVal s As String) As String
			Dim [rem] = i Mod 26
            s = CChar(ChrW(AscW("A"c) + [rem])) & s
			i = i \ 26 - 1
			Return If(i < 0, s, GetAlphaColumnHeader(i, s))
		End Function

		' create a polygon based on a list of points
		Private Shared Function CreatePolygon(ParamArray ByVal values() As Double) As Polygon
			Dim p = New Polygon()
			p.Points = New PointCollection()
			For i As Integer = 0 To values.Length - 2 Step 2
				p.Points.Add(New Point(values(i), values(i + 1)))
			Next i
			Return p
		End Function
	End Class
End Namespace

File: NET_4.5.2/C1.WPF.FlexGrid/CS/ExcelGrid/ExcelGrid/FlexExcel/ExcelCellFactory.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Shapes;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using System.Collections.Generic;
using System.Text;
using C1.WPF.FlexGrid;

namespace ExcelGrid
{
    /// <summary>
    /// Custom cell factory that displays Excel-style row and column headers.
    /// </summary>
    public class ExcelCellFactory : CellFactory
    {
        public override void CreateTopLeftContent(C1FlexGrid grid, Border bdr, CellRange range)
        {
            var poly = CreatePolygon(0, 10, 10, 10, 10, 0);
            poly.Fill = new SolidColorBrush(Color.FromArgb(0xe0, 0xf5, 0xf5, 0xf5));
            poly.VerticalAlignment = VerticalAlignment.Center;
            poly.HorizontalAlignment = HorizontalAlignment.Right;
            poly.Margin = new Thickness(2);
            bdr.Child = poly;
            bdr.Tag = grid;
            bdr.MouseLeftButtonDown += new MouseButtonEventHandler(bdr_MouseLeftButtonDown);
        }
        void bdr_MouseLeftButtonDown(object sender, MouseButtonEventArgs e)
        {
            // select the whole grid when user clicks the top left cell
            var g = ((FrameworkElement)sender).Tag as C1FlexGrid;
            g.SelectAll();
        }
        public override void CreateRowHeaderContent(C1FlexGrid grid, Border bdr, CellRange range)
        {
            var tb = new TextBlock();
            tb.HorizontalAlignment = HorizontalAlignment.Center;
            tb.VerticalAlignment = VerticalAlignment.Center;
            tb.Text = string.Format("{0}", range.Row + 1);
            bdr.Child = tb;
        }
        public override void CreateColumnHeaderContent(C1FlexGrid grid, Border bdr, CellRange range)
        {
            var tb = new TextBlock();
            tb.HorizontalAlignment = HorizontalAlignment.Center;
            tb.VerticalAlignment = VerticalAlignment.Center;
            tb.Text = GetAlphaColumnHeader(range.Column);
            bdr.Child = tb;
        }

        // create Excel style column headers ("A", "B", ... "AB", etc)
        string GetAlphaColumnHeader(int i)
        {
            return GetAlphaColumnHeader(i, string.Empty);
        }
        string GetAlphaColumnHeader(int i, string s)
        {
            var rem = i % 26;
            s = (char)((int)'A' + rem) + s;
            i = i / 26 - 1;
            return i < 0 ? s : GetAlphaColumnHeader(i, s);
        }

        // create a polygon based on a list of points
        static Polygon CreatePolygon(params double[] values)
        {
            var p = new Polygon();
            p.Points = new PointCollection();
            for (int i = 0; i < values.Length - 1; i += 2)
            {
                p.Points.Add(new Point(values[i], values[i + 1]));
            }
            return p;
        }
    }
}
